bundle/deploy/terraform/state_pull.go

```diff
@@ -24,71 +28,95 @@ func (l *statePull) Name() string {
-	f, err := l.filerFactory(b)
+	state := &tfState{}
+	err = json.Unmarshal(content, state)
 	if err != nil {
-		return diag.FromErr(err)
+		return nil, nil, err
 	}
 
+	return state, content, nil
+}
+
+func (l *statePull) localState(ctx context.Context, b *bundle.Bundle) (*tfState, error) {
 	dir, err := Dir(ctx, b)
 	if err != nil {
-		return diag.FromErr(err)
+		return nil, err
 	}
 
-	// Download state file from filer to local cache directory.
-	log.Infof(ctx, "Opening remote state file")
-	remote, err := l.remoteState(ctx, f)
+	content, err := os.ReadFile(filepath.Join(dir, TerraformStateFileName))
 	if err != nil {
-		log.Infof(ctx, "Unable to open remote state file: %s", err)
-		return diag.FromErr(err)
+		return nil, err
 	}
-	if remote == nil {
-		log.Infof(ctx, "Remote state file does not exist")
-		return nil
+
+	state := &tfState{}
+	err = json.Unmarshal(content, state)
+	if err != nil {
+		return nil, err
 	}
 
-	// Expect the state file to live under dir.
-	local, err := os.OpenFile(filepath.Join(dir, TerraformStateFileName), os.O_CREATE|os.O_RDWR, 0600)
+	return state, nil
+}
+
+func (l *statePull) Apply(ctx context.Context, b *bundle.Bundle) diag.Diagnostics {
+	dir, err := Dir(ctx, b)
 	if err != nil {
 		return diag.FromErr(err)
 	}
-	defer local.Close()
 
-	if !IsLocalStateStale(local, bytes.NewReader(remote.Bytes())) {
-		log.Infof(ctx, "Local state is the same or newer, ignoring remote state")
+	localStatePath := filepath.Join(dir, TerraformStateFileName)
+
+	// Case: remote state file does not exist. In this case we should not use the
+	// local state file because we cannot guarantee it corresponds to the same deployment,
+	// that is the same root_path and workspace host.
+	remoteState, remoteContent, err := l.remoteState(ctx, b)
+	if errors.Is(err, fs.ErrNotExist) {
+		log.Infof(ctx, "Remote state file does not exist. Invalidating local terraform state.")
+		os.Remove(localStatePath)
 		return nil
 	}
+	if err != nil {
+		return diag.Errorf("failed to read remote state file: %v", err)
+	}
 
-	// Truncating the file before writing
-	local.Truncate(0)
-	local.Seek(0, 0)
-
-	// Write file to disk.
-	log.Infof(ctx, "Writing remote state file to local cache directory")
-	_, err = io.Copy(local, bytes.NewReader(remote.Bytes()))
+	// Case: Local state file does not exist. In this case we should rely on the remote state file.
+	localState, err := l.localState(ctx, b)
+	if errors.Is(err, fs.ErrNotExist) {
+		log.Infof(ctx, "Local state file does not exist. Using remote terraform state. Invalidating local terraform state.")
+		err := os.WriteFile(localStatePath, remoteContent, 0600)
+		return diag.FromErr(err)
+	}
 	if err != nil {
+		return diag.Errorf("failed to read local state file: %v", err)
+	}
+
+	// If the lineages do not match, the terraform state files do not correspond to the same deployment.
+	if localState.Lineage != remoteState.Lineage {
```

Comment:
Where did the local one come from? This might happen if you swap out the root path or the workspace after doing a successful deploy and the new root path in the new workspace has an existing deployment.

Reply:
Exactly. Local states can persist even when we change `root_path` or host.